### ISSUE TITLE
Proposal: Allow omitting `partial` for a single definition

### ISSUE THREAD

**Issue by gulshan** [dotnet/csharplang#950]
Now keyword `partial` is required for all definitions of a class. Allow to omit the keyword for a single definition. It means, if there are n partial definitions of a class, allow a single of those one definitions to omit the partial keyword, while for other n-1 definitions partial keyword will still be required. This is just a cosmetic change. But omitting partial for a definition will help the developers to presume it is the main definition, while others are supportive. This will also encourage to handle secondary aspects of a class definition (like implementing an interface) in a partial definition. The code will reflect the separation of concern. While all of this is possible now, as partial is required for all definitions, one would rather add to the main definition rather than making it partial. This proposal is influenced by swift, where making an extension to class to handle secondary concerns (even in the same file) is a common practice.

**Comment by alrz** [comment]
Relates to https://github.com/dotnet/roslyn/issues/6953

**Comment by gulshan** [comment]
It seems the issue with the roslyn issues was that, it makes partial totally optional making the partial keyword kind of irrelevant. It would enable accidental merging of classes instead of showing duplication error. This proposal mostly avoids that. If it were not a breaking change, I would propose marking one single class without partial keyword mandatory for the the full group of partial classes (or partial definitions of a single class). That would prevent the case in https://github.com/dotnet/roslyn/issues/6953#issuecomment-159048208 . That is a rare case though, as far as my experience goes.

Pinging @mattwar as he was the champion for the roslyn issue.

**Comment by alrz** [comment]
>It seems the issue with the roslyn issues was that, it makes partial totally optional 

No, that was proposed in the opposite of this issue: only one of type decls needs to be partial, like VB. Allowing the omission of partial only for a single decl might be actually preferable and address some of concerns regarding accidental type merges.

**Comment by jnm2** [comment]
Not sure I like this. I like being able to see a single file and know conclusively if there are other parts of the class. Also reminds me of how the LDT wishes they had made an `iterator` modifier mandatory like they later did for `async`.

**Comment by alrz** [comment]
@jnm2 

This was proposal alongside of generators, so you don't need to mark types as `partial` in the source, if a generator is generating other parts for you under the hood. Just a piece of information that you might not know at the time. see https://github.com/dotnet/roslyn/issues/6953#issuecomment-186382974

**Comment by gulshan** [comment]
One thing came to my mind regarding `async` modifier enforcement. Now it's easy to avoid specifying `async` on an asynchronous method. Just make a local function with async and return it. As we can await any task regardless of being marked async, omission of async from public API are not that much of a problem in my opinion. Something returning task is enough to determine it can be called asynchronously.

**Comment by alrz** [comment]
async is not a part of public api, if you're going to use await, you need to use async, for disambiguation in particular cases. it's insignificant from call-site's point of view.

**Comment by yaakov-h** [comment]
I agree with @jnm2 - I think that `partial` should remain explicit for all user-defined partial class definitions. Unfortunately, generators complicate this. 😞 

Any way we could have both?

@gulshan Extension Everything (#192 / https://github.com/dotnet/roslyn/issues/11159) is more akin to Swift extensions than partial classes.

**Comment by gulshan** [comment]
My point is, async is more of an convenient return value packer-transformer(await is the unpacker-transformer) rather than mark of a method being asynchronous.

`partial` is already absent in public api. And the scope of partial definition is limited to a namespace. It will create a little bit of pressure on a developer to always consider there can be a partial definition somewhere within the current namespace. But extension methods already create that pressure in my opinion, with no limit to the scope. But allowing to omit partial will encourage developers to organize code with separation of concern. At present, developers are using partial for the verbosity I think. And the proposed extension everything is also going to be verbose making developers reluctant to use them.

**Comment by DavidArno** [comment]
> But omitting partial for a definition will help the developers to presume it is the main definition

I disagree. It will simply mislead developers into thinking the class isn't partial if they look at the "wrong" file. This is classic "pit of success" stuff. I'd go beyond @jnm2's "not sure I like this" position to arguing this change would be irresponsible.

**Comment by DavidArno** [comment]
> My point is, async is more of an convenient return value packer-transformer(await is the unpacker-transformer) rather than mark of a method being asynchronous.

No. It exits *only* to clarify the meaning of `await` within a method. It's pure noise, which is unfortunately needed to avoid breaking existing code.

**Comment by alrz** [comment]
>It's pure noise, which is unfortunately needed to avoid breaking existing code.

It turned out that it's actually useful since an async without await produces a warning, you will know that any async method is going to be rewritten as a state machine. As stephentoub pointed out that in itself might be a cause of a bug etc. Which leads us to `iterator` modifier being proposed in the future which is going to be a code conversion choice for existing methods and mandatory for async iterators.

Though I stil don't understand how it's relevant here.

**Comment by AtsushiKan** [comment]
>It will create a little bit of pressure on a developer to always consider there can be a partial definition somewhere within the current namespace. But extension methods already create that pressure in my opinion

It's not the same thing. Extension methods can only build upon the type's external surface area just like another user of that type. Being able to know there are no other files with dependencies on private class members and implementation details is still useful.
